lib/debug/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+use cranelift_codegen::isa::TargetFrontendConfig;
+use faerie::{Artifact, Decl};
+use std::fmt;
+use target_lexicon::{BinaryFormat, Triple};
+
+pub use crate::read_debuginfo::{read_debuginfo, DebugInfoData};
+pub use crate::transform::transform_dwarf;
+pub use crate::write_debuginfo::{emit_dwarf, ResolvedSymbol, SymbolResolver};
+
+use wasmtime_environ::AddressTransforms;
+
+mod read_debuginfo;
+mod transform;
+mod write_debuginfo;
+
+impl fmt::Display for transform::Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Transform error: ")?;
+        match self {
+            transform::Error::GimliError(e) => write!(f, "gimli error, {}", e),
+            transform::Error::Custom(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+struct FunctionRelocResolver {}
+impl SymbolResolver for FunctionRelocResolver {
+    fn resolve_symbol(&self, symbol: usize, addend: i64) -> ResolvedSymbol {
+        let name = format!("_wasm_function_{}", symbol);
+        ResolvedSymbol::Reloc { name, addend }
+    }
+}
+
+pub fn emit_debugsections(
+    obj: &mut Artifact,
+    target_config: &TargetFrontendConfig,
+    debuginfo_data: &DebugInfoData,
+    at: &AddressTransforms,
+) -> Result<(), String> {
+    let dwarf =
+        transform_dwarf(target_config, debuginfo_data, at).map_err(|err| format!("{}", err))?;
+    let resolver = FunctionRelocResolver {};
+    emit_dwarf(obj, &dwarf, &resolver);
+    Ok(())
+}
+
+struct ImageRelocResolver<'a> {
+    func_offsets: &'a Vec<u64>,
+}
+
+impl<'a> SymbolResolver for ImageRelocResolver<'a> {
+    fn resolve_symbol(&self, symbol: usize, addend: i64) -> ResolvedSymbol {
+        let func_start = self.func_offsets[symbol];
+        ResolvedSymbol::PhysicalAddress(func_start + addend as u64)
+    }
+}
+
+pub fn emit_debugsections_image(
+    triple: Triple,
+    target_config: &TargetFrontendConfig,
+    debuginfo_data: &DebugInfoData,
+    at: &AddressTransforms,
+    funcs: &Vec<(*const u8, &[u8])>,
+) -> Result<Vec<u8>, String> {
+    let ref func_offsets = funcs
+        .iter()
+        .map(|(ptr, _)| *ptr as u64)
+        .collect::<Vec<u64>>();
+    let mut obj = Artifact::new(triple, String::from("module"));
+    let dwarf =
+        transform_dwarf(target_config, debuginfo_data, at).map_err(|err| format!("{}", err))?;
+    let resolver = ImageRelocResolver { func_offsets };
+    for (i, (_, body)) in funcs.iter().enumerate() {
+        let name = format!("___wasm_function_{}", i);
+        obj.declare_with(&name, Decl::Function { global: false }, body.to_vec())
+            .map_err(|_| String::from("Unable to defaine a function"))?;
+    }
+    emit_dwarf(&mut obj, &dwarf, &resolver);
+
+    // LLDB is too "magical" about mach-o, generating elf
+    let mut bytes = obj
+        .emit_as(BinaryFormat::Elf)
+        .map_err(|_| String::from("Unable to emit"))?;
+    // elf is still missing details...
+    patch_faerie_elf(&mut bytes, funcs);
```

Comment:
Can we fix faerie instead?

Reply:
It needs to be refactored in something like "light linker" to produce loadable object file. No sure if it is in scope of the faerie.